twisted-server-1/transformedpoetry.py

```diff
@@ -45,20 +45,37 @@ def cummingsify(self, poem):
 
 
 class TransformProtocol(NetstringReceiver):
+    '''The goal of separating the protocol and service ideally would put the
+    utf8 processing in the service. However python names are now unicode.
+    That means we can not use a byte string in the getattr call.
+    
+    I put the decode into stringReceived and overrode sendString to encode the data.
+    
+    Ideally we would probably want the decode/encode in the TransformService.
+    
+    The error handling was modified to provide feedback about what was truly failing.
+    The original except returning None was not helpful.
+    '''
 
     def stringReceived(self, request):
-        if '.' not in request: # bad request
+        if b'.' not in request: # bad request
             self.transport.loseConnection()
             return
 
-        xform_name, poem = request.split('.', 1)
+        xform_name, poem = request.decode('utf8').split('.', 1)
 
         self.xformRequestReceived(xform_name, poem)
 
-    def xformRequestReceived(self, xform_name, poem):
-        new_poem = self.factory.transform(xform_name, poem)
+    def sendString(self, string):
+        # Netstrings are really byte strings
+        super().sendString(string.encode('utf8'))
 
-        if new_poem is not None:
+    def xformRequestReceived(self, xform_name, poem):
+        try:
+            new_poem = self.factory.transform(xform_name, poem)
+        except ValueError as exc:
+            self.sendString(str(exc))
```

Comment:
I'm not sure I agree with this change. Errors should be clearly distinguished from correct return values. This silently returns an error as if it were the result of the transform.

Reply:
OK. I'll review my changes and find a better solution. Some of the original error handling was a bit too optimistic since now even a print statement is fraught with the risk of a unicode error. The original ASCII poems were risk free in Python 2.

Reply:
I pushed the changes.

The server now logs errors that had been silently swallowed before. The protocol works as specified; only poems are sent.